addons/web/static/tests/boot_tests.js

```diff
@@ -1,18 +1,66 @@
-(function () {
+(async function () {
 "use strict";
 
-odoo.__DEBUG__.didLogInfo.then(function () {
-    var modulesInfo = odoo.__DEBUG__.jsModules;
+    // We define here a test asserting that no module failed to be loaded. The
+    // module system (see boot.js) waits 10 seconds before logging that some
+    // modules couldn't be loaded, when it happens. The test below is really
+    // important as it will detect testing modules that can't be loaded, for
+    // instance because one of their dependencies is missing. Without it, we
+    // could have tests that aren't executed, without any notice.
+    //
+    // We have the following constraints around this test:
+    //  - we want it to be always executed (obviously)
+    //  - we need it to wait for the module system to complete the loading of
+    //    the modules before being executed (wait for the didLogInfo promise)
+    //  - we don't want it to slow down the whole suite
+    //  - we can't rely on the module system to define this test, as it is
+    //    partially testing it as well
+    //
+    // To achieve this, we ensure that this test is defined after (some of) the
+    // other tests. Almost all tests need module 'web.test_utils', which is
+    // asynchronous (it loads resources that are lazy-loaded outside the test
+    // environment). So here, we also wait for 'web.test_utils' (without using
+    // the module system) before defining our test, so that our test is defined
+    // after all those requiring 'web.test_utils'. That way:
+    //  - either the suite takes more than 10 seconds to execute, and once we
+    //    reach our test, the promise is already resolved and it can directly
+    //    check if some modules failed
+    //  - or the suite takes less than 10 seconds, and our test will wait for
+    //    the promise to be resolved.
+    // However, in any cases, it won't slow down the test suite longer than
+    // necessary.
+    let resolve;
+    const prom = new Promise(r => {
+        resolve = r;
+    });
+    const timeout = setTimeout(() => {
+        // module 'web.test_utils' must be loaded after a while, otherwise it
+        // could mean it has been renamed, and our test will never been executed
+        if (!odoo.__DEBUG__.services['web.test_utils']) {
+            throw new Error('Module "web.test_utils" could not be loaded');
+        }
+    }, 5000);
+    const interv = setInterval(() => {
+        if (odoo.__DEBUG__.services['web.test_utils']) {
+            clearInterval(interv);
+            clearTimeout(timeout);
+            setTimeout(resolve, 0);
+        }
+    }, 50);
+    await prom;
```

Comment:
Isn't there still the possibility to have the same issue as before since you wait on something before registering the test in QUnit?
At least during the 50ms frame of the interval. And that is the most optimistic case, where the logic based `web.test_utils` is true, if the module is indeed loaded (because nobody will know about this condition when writing a test, and it can happen that it is not used).

Reply:
you could, on a suite that doesn't use test_utils. We don't have the case in Odoo as far as i know

Reply:
I'm still not sure we have the right approach here.

I feel like the `test successful` in `QUnit.done` should never be sent before `didLogInfo` is resolved.

Reply:
Critical info I just found: `QUnit.start();` is actually called in `web.test_utils` so there is actually no way with the current code that module is not required in a test. 

I don't think it's a good idea to start the tests there at all, and it should be linked to some kind of "boot" instead, but that seems very convenient for our fix/hack here.

Reply:
Apparently the goal is to call `start` after all tests have been registered, it seems natural that even starting the tests should depend on `didLogInfo`.

Reply:
I disagree about your very last statement: we don't want to wait for didLogInfo before starting the tests, for the reasons I already exposed. However I like your idea of waiting for didLogInfo in QUnit.done, before printing 'test successful'. That may solve the issue. I'll give it a try. There is another constraint I did not mention earlier: when you run the tests in the browser, you must see that the suite fails when there is a failing module.